tests/jballerina-unit-test/src/test/resources/test-src/expressions/varref/record-variable-reference-semantics-negative.bal

```diff
@@ -198,3 +198,23 @@ function testFieldAndIndexBasedVarRefs() returns [anydata, anydata] {
     {name: m["var1"], yearAndAge: [m["var2"], _]} = ch3;
     return [m["var1"], m["var2"]];
 }
+
+type Employee record {
+    string name;
+    int id;
+    int age?;
+};
+
+function testOptionalFieldsInRecordBindingPattern(){
+   Employee e = {name: "Jo", id: 1234};
+
+   string eName;
+   int eId;
+   int eAge;
+
+   {name: eName, id: eId, age: eAge} = e;
+
+   var {name: name, id: id, age: age} = e;
```

Comment:
For all kinds, let's add some tests for when the field is just the variable-name also, and not `field-name : binding-pattern`.

```ballerina
var {name, id: id, age}
```

Reply:
Also for when the field binding patterns are within other binding patterns.

e.g., for
```ballerina
type Employee record {
    string name;
    record {
        int id;
        int age?;
    } details?;
};
```

and

```ballerina
type Employee record {
    string name;
    record {
        int id;
        int age?;
    } details;
};
```

Reply:
And part of list and error binding patterns.

This doesn't fail atm.

```ballerina
type Employee record {
    string name;
    record {
        int id;
        int age?;
    } details?;
};

function fn() {
    Employee[2] e = [{name: "Jo", details: {id: 1234}}, {name: "Amy", details: {id: 1234}}];

    var [{name: nameNew, details: {id: idNew, age: ageNew}}, ...x] = e;
}
```

Reply:
Fixed this. But to handle the recursive records had to introduce two methods because data types are different in destructure expression and var expression